File: java/src/main/java/com/sweetmanor/algorithm/csp/CSP.java
package com.sweetmanor.algorithm.csp;

import java.util.*;

/**
 * 约束满足问题 —— 《算法精粹》
 * <p>
 * V - 变量类型 ， D - 值域类型
 *
 * @author ijlhjj
 * @version 1.0 2023-12-13
 */
public class CSP<V, D> {

    private List<V> variables; //变量
    private Map<V, List<D>> domains; //值域
    private Map<V, List<Constraint<V, D>>> constraints = new HashMap<>();//约束

    public CSP(List<V> variables, Map<V, List<D>> domains) {
        this.variables = variables;
        this.domains = domains;

        //每个变量默认添加一个空的约束列表
        for (V v : variables) {
            //校验变量必须在值域字典中
            if (!domains.containsKey(v))
                throw new IllegalArgumentException("Every variable should have a domain assigned to it.");
            constraints.put(v, new ArrayList<>());
        }
    }

    /**
     * 增加约束
     */
    public void addConstraint(Constraint<V, D> constraint) {
        for (V v : constraint.variables) {
            //新增约束的变量必须包含在当前变量列表中
            if (!variables.contains(v))
                throw new IllegalArgumentException("Variable in constraint not in CSP");
            constraints.get(v).add(constraint); //约束添加到约束字典
        }
    }

    /**
     * 检查当前赋值是否满足所有约束
     */
    public boolean consistent(V v, Map<V, D> assignment) {
        for (Constraint<V, D> constraint : constraints.get(v)) {
            //校验是否满足约束条件
            if (!constraint.satisfied(assignment))
                return false;
        }
        return true;
    }

    /**
     * 递归回溯搜索
     */
    public Map<V, D> backtrackingSearch(Map<V, D> assignment) {
        //所有变量已赋值，完成搜索
        if (assignment.size() == variables.size())
            return assignment;

        //找一个未赋值的变量
        V unassigned = variables.stream().filter(v -> !assignment.containsKey(v)).findFirst().get();

        //尝试变量所有可能赋值
        for (D value : domains.get(unassigned)) {
            //创建副本，并增加当前赋值
            Map<V, D> localAssignment = new HashMap<>(assignment);
            localAssignment.put(unassigned, value);

            //检查当前赋值是否满足约束条件
            if (consistent(unassigned, localAssignment)) {
                Map<V, D> result = backtrackingSearch(localAssignment); //递归求解

                //找到满足约束的解
                if (result != null)
                    return result;
            }
        }

        //没有找到解时返回 null
        return Collections.emptyMap();
    }

    /**
     * 回溯搜索
     */
    public Map<V, D> backtrackingSearch() {
        return backtrackingSearch(new HashMap<>());
    }

}

File: python/src/com/sweetmanor/algorithm/csp/csp.py
# 《算法精粹》
# 约束满足问题

from abc import ABC, abstractmethod
from typing import TypeVar, Generic, List, Dict, Optional

V = TypeVar('V')
D = TypeVar('D')


class Constraint(Generic[V, D], ABC):
    def __init__(self, variables: List[V]) -> None:
        self.variables = variables  # 变量

    @abstractmethod
    def satisfied(self, assignment: Dict[V, D]) -> bool:
        """ 校验是否满足约束条件 """
        ...


class CSP(Generic[V, D]):

    def __init__(self, variables: List[V], domains: Dict[V, List[D]]) -> None:
        self.variables = variables  # 变量
        self.domains = domains  # 值域
        self.constraints: Dict[V, List[Constraint[V, D]]] = {}  # 约束

        # 每个变量默认添加一个空的约束列表
        for var in self.variables:
            # 校验变量必须在值域字典中
            if var not in self.domains:
                raise LookupError("Every variable should have a domain assigned to it.")
            self.constraints[var] = []

    def add_constraint(self, constraint: Constraint[V, D]) -> None:
        """ 增加约束 """
        for var in constraint.variables:
            # 新增约束的变量必须包含在当前变量列表中
            if var not in self.variables:
                raise LookupError("Variable in constraint not in CSP")
            self.constraints[var].append(constraint)  # 约束添加到约束字典

    def consistent(self, variable: V, assignment: Dict[V, D]) -> bool:
        """ 检查当前赋值是否满足所有约束 """
        for constraint in self.constraints[variable]:
            # 校验是否满足约束条件
            if not constraint.satisfied(assignment):
                return False
        return True

    def backtracking_search(self, assignment: Dict[V, D] = {}) -> Optional[Dict[V, D]]:
        """ 递归回溯搜索 """

        # 所有变量已赋值，完成搜索
        if len(assignment) == len(self.variables):
            return assignment

        # 找一个未赋值的变量
        unassigned = [v for v in self.variables if v not in assignment]
        first = unassigned[0]

        # 尝试变量所有可能赋值
        for val in self.domains[first]:
            # 创建副本，并增加当前赋值
            local_assignment = assignment.copy()
            local_assignment[first] = val

            # 检查当前赋值是否满足约束条件
            if self.consistent(first, local_assignment):
                result = self.backtracking_search(local_assignment)  # 递归求解
                # 找到满足约束的解
                if result is not None:
                    return result

        # 没有找到解时返回 None
        return None
